PhysicsTools/PatAlgos/plugins/PATTracksToPackedCandidates.cc

```diff
@@ -0,0 +1,203 @@
+void PATTracksToPackedCandidates::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  using namespace std;
+
+  //track collection
+  auto tracks = iEvent.getHandle(srcTracks_);
+
+  auto outPtrTrksAsCands = std::make_unique<std::vector<pat::PackedCandidate>>();
+
+  //vtx collection
+  edm::Handle<reco::VertexCollection> pvs;
+  iEvent.getByToken(srcPrimaryVertices_, pvs);
+  reco::VertexRef pv(pvs.id());
+  reco::VertexRefProd pvRefProd(pvs);
+  if (!pvs->empty()) {
+    pv = reco::VertexRef(pvs, 0);
+  }
+
+  //best vertex
+  double bestvzError = -999.9, bestvxError = -999.9, bestvyError = -999.9;
+  const reco::Vertex& vtx = (*pvs)[0];
+  bestvzError = vtx.zError();
+  bestvxError = vtx.xError();
+  bestvyError = vtx.yError();
+  const math::XYZPoint& bestvtx(vtx.position());
+  math::Error<3>::type vtx_cov = vtx.covariance();
+
+  std::vector<int> mapping(tracks->size(), -1);
+  int savedCandIndx = 0;
+  int trkIndx = -1;
+  for (auto const& trk : *tracks) {
+    trkIndx++;
+    double dzvtx = std::abs(trk.dz(bestvtx));
+    double dxyvtx = std::abs(trk.dxy(bestvtx));
+    double dzerror = std::hypot(trk.dzError(), bestvzError);
+    double dxyerror = sqrt(trk.dxyError(bestvtx, vtx_cov) * trk.dxyError(bestvtx, vtx_cov) + bestvxError * bestvyError);
+
+    if (dzvtx >= dzSigCut_ * dzerror)
+      continue;
+    if (dxyvtx >= dxySigCut_ * dxyerror)
+      continue;
+    if (trk.pt() >= ptMax_ || trk.pt() <= ptMin_)
+      continue;
+
+    bool passSelection = false;
+    if (dzvtx < dzSigHP_ * dzerror && dxyvtx < dxySigHP_ * dxyerror)
+      passSelection = true;
+
+    addPackedCandidate(*outPtrTrksAsCands, trk, pv, pvRefProd, passSelection);
+
+    //for creating the reco::Track -> pat::PackedCandidate map
+    mapping[trkIndx] = savedCandIndx;
+    savedCandIndx++;
+  }
+  edm::OrphanHandle<pat::PackedCandidateCollection> oh = iEvent.put(std::move(outPtrTrksAsCands));
+  auto tk2pc = std::make_unique<edm::Association<pat::PackedCandidateCollection>>(oh);
+  edm::Association<pat::PackedCandidateCollection>::Filler tk2pcFiller(*tk2pc);
+  tk2pcFiller.insert(tracks, mapping.begin(), mapping.end());
+  tk2pcFiller.fill();
+  iEvent.put(std::move(tk2pc));
+}
+
+void PATTracksToPackedCandidates::addPackedCandidate(std::vector<pat::PackedCandidate>& cands,
+                                                     const reco::Track trk,
+                                                     const reco::VertexRef& pvSlimmed,
+                                                     const reco::VertexRefProd& pvSlimmedColl,
+                                                     bool passPixelTrackSel) const {
+  const float mass = 0.13957018;
+
+  int id = 211 * trk.charge();
+
+  reco::Candidate::PolarLorentzVector p4(trk.pt(), trk.eta(), trk.phi(), mass);
+  cands.emplace_back(
+      pat::PackedCandidate(p4, trk.vertex(), trk.pt(), trk.eta(), trk.phi(), id, pvSlimmedColl, pvSlimmed.key()));
```

Comment:
```suggestion
  cands.emplace_back(p4, trk.vertex(), trk.pt(), trk.eta(), trk.phi(), id, pvSlimmedColl, pvSlimmed.key());
```
explicit type is not needed with emplace_back

Reply:
This comment is included in this commit: https://github.com/cms-sw/cmssw/pull/30313/commits/6e7159380e61d7f6bf7342a1f71162fbcace502a